File: chore_manager/commands/assign.py
import random
from typing import Any, Dict, List, Optional

from ..animations.spinner import show_title, spin_wheel
from ..utils.console import get_console
from ..utils.date_helper import get_today_date
from ..utils.env import should_animate
from ..utils.file_manager import (
    ASSIGNMENTS_PATH,
    CONFIG_PATH,
    HISTORY_PATH,
    TOTEM_PATH,
    load_json,
    save_json,
)
from ..utils.gradient import pastel_text, rainbow_text
from ..utils.totem_manager import normalize_totem
from ..utils.user_selector import select_user_for_chore


def _has_assignment_today(assignments: List[Dict[str, Any]], today: str) -> bool:
    return any(assignment.get("date") == today for assignment in assignments)


def assign_chores(
    auto_mode: bool = False,
    count: Optional[Any] = None,
    specific_chore: Optional[str] = None,
    force_mode: bool = False,
    animate: Optional[bool] = None,
    no_anim: bool = False,
) -> Dict[str, Any]:
    console = get_console()
    config = load_json(CONFIG_PATH)
    history = load_json(HISTORY_PATH)
    assignments = load_json(ASSIGNMENTS_PATH)

    today = get_today_date()

    if auto_mode and _has_assignment_today(assignments, today) and not force_mode:
        console.print("[yellow]⚠️  Les corvées ont déjà été attribuées aujourd'hui (mode auto).[/]")
        return {"skipped": True, "reason": "already_assigned", "date": today}

    if animate is None:
        animate = should_animate(no_anim=no_anim)

    show_title()

    totem_raw = load_json(TOTEM_PATH)
    normalized = normalize_totem(totem_raw, list(history.keys()))
    immune = normalized["immune"]
    forced_queue = normalized["forcedQueue"]
    safe = normalized["safe"]
    safe_set = set(safe)

    chores_to_assign: List[str] = []

    if specific_chore and count is not None:
        matching = [
            chore
            for chore in config["chores"]
            if specific_chore.lower() in chore.lower()
        ]

        if not matching:
            console.print(f"\n[red]❌ Corvée \"{specific_chore}\" non trouvée dans la configuration.\n[/]")
            console.print("[grey70]Corvées disponibles:[/]")
            for chore in config["chores"]:
                console.print(f"[grey70]  • {chore}[/]")
            console.print("")
            return {"error": "chore_not_found", "date": today}

        chore_to_assign = matching[0]
        try:
            num_count = int(count)
        except Exception:
            num_count = -1

        if num_count < 1:
            console.print("\n[red]❌ Le nombre doit être un nombre positif.\n[/]")
            return {"error": "invalid_count", "date": today}

        chores_to_assign = [chore_to_assign] * num_count

        console.print(f"[grey70]📅 Date : {today}[/]")
        console.print(f"[grey70]🎯 Corvée : {chore_to_assign} x{num_count}[/]")
        console.print("")

    elif count is not None and not specific_chore:
        try:
            num_count = int(count)
        except Exception:
            num_count = -1
        if num_count < 1:
            console.print("\n[red]❌ Le nombre de corvées doit être un nombre positif.\n[/]")
            return {"error": "invalid_count", "date": today}

        if num_count > len(config["chores"]):
            console.print(
                f"\n[yellow]⚠️  Vous avez demandé {num_count} corvées mais il n'y en a que {len(config['chores'])} disponibles.[/]"
            )
            console.print(
                f"[grey70]Toutes les {len(config['chores'])} corvées seront attribuées.\n[/]"
            )
            chores_to_assign = list(config["chores"])
        else:
            chores_to_assign = random.sample(config["chores"], num_count)

        if not chores_to_assign:
            chores_to_assign = list(config["chores"])

        console.print(f"[grey70]📅 Date : {today}[/]")
        console.print(
            f"[grey70]🎯 Nombre de corvées : {len(chores_to_assign)}/{len(config['chores'])}[/]"
        )
        console.print("")

    else:
        chores_to_assign = list(config["chores"])
        console.print(f"[grey70]📅 Date : {today}[/]")
        console.print(f"[grey70]🎯 Toutes les corvées ({len(chores_to_assign)})[/]")
        console.print("")

    users = list(history.keys())
    non_safe_users = [user for user in users if user not in safe_set]
    if non_safe_users:
        total = sum(history[user] for user in non_safe_users)
        avg = round(total / len(non_safe_users))
        for user in safe:
            history[user] = avg

    eligible_users = [user for user in users if user not in safe_set]
    if not eligible_users:
        console.print("\n[red]❌ Aucun utilisateur disponible : tout le monde est \"à l'abri\".\n[/]")
        return {"error": "no_eligible_users", "date": today}

    def show_safe_art(user: str) -> None:
        console.print(
            "[cyan]   .-\"\"\"\"-.\n  /  _  _  \\\n |  (o)(o)  |\n |   .__.   |\n  \\  ----  /\n   `-.__.-`[/]"
        )
        console.print(f"[cyan]🛡️ {user} est à l'abri. Tour annulé, on relance.[/]")

    def select_user_for_run(
        run_history: Dict[str, int],
        run_immune: List[str],
        run_forced_queue: List[str],
        ignore_safe: bool,
    ) -> Dict[str, Any]:
        if run_forced_queue:
            forced_index = next(
                (idx for idx, user in enumerate(run_forced_queue) if user not in safe_set),
                -1,
            )
            if forced_index != -1:
                forced_user = run_forced_queue.pop(forced_index)
                return {"selected_user": forced_user, "restart_all": False}
            if not ignore_safe and run_forced_queue and run_forced_queue[0] in safe_set:
                return {
                    "selected_user": run_forced_queue[0],
                    "restart_all": True,
                    "safe_hit": True,
                }

        exclude_safe = safe if ignore_safe else []
        candidate_all = select_user_for_chore(run_history, exclude=exclude_safe)
        if not candidate_all:
            return {"selected_user": None, "restart_all": False}

        if not ignore_safe and candidate_all in safe_set:
            return {"selected_user": candidate_all, "restart_all": True, "safe_hit": True}

        if candidate_all in run_immune:
            exclude_list = list(set(run_immune + safe))
            alternative = select_user_for_chore(run_history, exclude=exclude_list)
            if alternative:
                if candidate_all in run_immune:
                    run_immune.remove(candidate_all)
                if candidate_all not in run_forced_queue:
                    run_forced_queue.append(candidate_all)
                console.print(f"[grey70]🛡️ Totem actif : {candidate_all} reporté au prochain tour.[/]")
                return {"selected_user": alternative, "restart_all": False}

        return {"selected_user": candidate_all, "restart_all": False}

    today_assignments: List[Dict[str, Any]] = []
    ignore_safe = False
    attempts = 0
    max_attempts = 5

    while attempts < max_attempts:
        attempts += 1
        run_history = dict(history)
        run_assignments: List[Dict[str, Any]] = []
        run_immune = list(immune)
        run_forced_queue = list(forced_queue)
        restart_all = False

        for chore in chores_to_assign:
            result = select_user_for_run(run_history, run_immune, run_forced_queue, ignore_safe)
            if result.get("restart_all"):
                if result.get("safe_hit") and result.get("selected_user"):
                    spin_wheel(users, result["selected_user"], chore, animate=animate)
                    show_safe_art(result["selected_user"])
                restart_all = True
                ignore_safe = True
                break

            selected_user = result.get("selected_user")
            if not selected_user:
                console.print("\n[red]❌ Impossible de sélectionner un utilisateur non protégé.\n[/]")
                return {"error": "no_candidate", "date": today}

            run_history[selected_user] += 1
            run_assignments.append(
                {
                    "date": today,
                    "chore": chore,
                    "user": selected_user,
                    "completed": False,
                }
            )

        if restart_all:
            continue

        for assignment in run_assignments:
            spin_wheel(users, assignment["user"], assignment["chore"], animate=animate)
            today_assignments.append(assignment)
            assignments.append(assignment)

        history.update(run_history)
        immune = run_immune
        forced_queue = run_forced_queue
        break

    if attempts >= max_attempts:
        console.print("\n[red]❌ Trop de relances dues aux utilisateurs \"à l'abri\".\n[/]")
        return {"error": "too_many_retries", "date": today}

    save_json(HISTORY_PATH, history)
    save_json(ASSIGNMENTS_PATH, assignments)
    save_json(TOTEM_PATH, {"immune": immune, "forcedQueue": forced_queue, "safe": safe})

    console.print("")
    console.print(rainbow_text("═" * 50))
    console.print(pastel_text("✨ Répartition terminée avec succès ! ✨"))
    console.print(rainbow_text("═" * 50))
    console.print("")

    return {
        "date": today,
        "assignments": today_assignments,
        "history": history,
        "totem": {"immune": immune, "forcedQueue": forced_queue, "safe": safe},
    }

File: legacy/node/src/commands/assign.js
const chalk = require('chalk');
const gradient = require('gradient-string');
const { loadJSON, saveJSON, CONFIG_PATH, HISTORY_PATH, ASSIGNMENTS_PATH, TOTEM_PATH } = require('../utils/fileManager');
const { getTodayDate } = require('../utils/dateHelper');
const { selectUserForChore } = require('../utils/userSelector');
const { normalizeTotem } = require('../utils/totemManager');
const { spinWheel, showTitle } = require('../animations/spinner');

// Vérifier si une attribution existe déjà aujourd'hui
function hasAssignmentToday(assignments) {
  const today = getTodayDate();
  return assignments.some(assignment => assignment.date === today);
}

// Commande: Répartir les corvées
async function assignChores(autoMode = false, count = null, specificChore = null, forceMode = false) {
  const config = loadJSON(CONFIG_PATH);
  const history = loadJSON(HISTORY_PATH);
  const assignments = loadJSON(ASSIGNMENTS_PATH);
  
  const today = getTodayDate();
  
  // Vérifier si des attributions existent déjà aujourd'hui
  if (autoMode && hasAssignmentToday(assignments) && !forceMode) {
    console.log(chalk.yellow('⚠️  Les corvées ont déjà été attribuées aujourd\'hui (mode auto).'));
    return;
  }
  
  // Afficher le titre animé
  showTitle();

  // Charger et normaliser le totem
  const totemRaw = loadJSON(TOTEM_PATH);
  let { immune, forcedQueue, safe } = normalizeTotem(totemRaw, Object.keys(history));
  const safeSet = new Set(safe);
  
  // Déterminer les corvées à attribuer
  let choresToAssign = [];
  
  // Cas 1: Corvée spécifique avec count
  if (specificChore && count !== null) {
    // Vérifier que la corvée existe
    const matchingChores = config.chores.filter(c => 
      c.toLowerCase().includes(specificChore.toLowerCase())
    );
    
    if (matchingChores.length === 0) {
      console.log(chalk.red(`\n❌ Corvée "${specificChore}" non trouvée dans la configuration.\n`));
      console.log(chalk.gray('Corvées disponibles:'));
      config.chores.forEach(c => console.log(chalk.gray(`  • ${c}`)));
      console.log('');
      return;
    }
    
    const choreToAssign = matchingChores[0];
    const numCount = parseInt(count);
    
    if (isNaN(numCount) || numCount < 1) {
      console.log(chalk.red('\n❌ Le nombre doit être un nombre positif.\n'));
      return;
    }
    
    // Créer un tableau avec la corvée répétée N fois
    choresToAssign = Array(numCount).fill(choreToAssign);
    
    console.log(chalk.gray(`📅 Date : ${today}`));
    console.log(chalk.gray(`🎯 Corvée : ${choreToAssign} x${numCount}`));
    console.log('');
    
  }
  // Cas 2: Count sans corvée spécifique (sélection aléatoire)
  else if (count !== null && !specificChore) {
    const numCount = parseInt(count);
    if (isNaN(numCount) || numCount < 1) {
      console.log(chalk.red('\n❌ Le nombre de corvées doit être un nombre positif.\n'));
      return;
    }
    if (numCount > config.chores.length) {
      console.log(chalk.yellow(`\n⚠️  Vous avez demandé ${numCount} corvées mais il n'y en a que ${config.chores.length} disponibles.`));
      console.log(chalk.gray(`Toutes les ${config.chores.length} corvées seront attribuées.\n`));
    } else {
      // Sélectionner aléatoirement les corvées
      const shuffled = [...config.chores].sort(() => Math.random() - 0.5);
      choresToAssign = shuffled.slice(0, numCount);
    }
    
    if (choresToAssign.length === 0) {
      choresToAssign = config.chores;
    }
    
    console.log(chalk.gray(`📅 Date : ${today}`));
    console.log(chalk.gray(`🎯 Nombre de corvées : ${choresToAssign.length}/${config.chores.length}`));
    console.log('');
  }
  // Cas 3: Toutes les corvées (défaut)
  else {
    choresToAssign = config.chores;
    console.log(chalk.gray(`📅 Date : ${today}`));
    console.log(chalk.gray(`🎯 Toutes les corvées (${choresToAssign.length})`));
    console.log('');
  }
  
  const todayAssignments = [];
  const users = Object.keys(history);
  const nonSafeUsers = users.filter(user => !safeSet.has(user));
  if (nonSafeUsers.length > 0) {
    const total = nonSafeUsers.reduce((sum, user) => sum + history[user], 0);
    const avg = Math.round(total / nonSafeUsers.length);
    safe.forEach(user => {
      history[user] = avg;
    });
  }
  const eligibleUsers = users.filter(user => !safeSet.has(user));

  if (eligibleUsers.length === 0) {
    console.log(chalk.red('\n❌ Aucun utilisateur disponible : tout le monde est "à l\'abri".\n'));
    return;
  }
  
  function showSafeArt(user) {
    console.log(chalk.cyan('   .-""""-.\n  /  _  _  \\\n |  (o)(o)  |\n |   .__.   |\n  \\  ----  /\n   `-.__.-`'));
    console.log(chalk.cyan(`🛡️ ${user} est à l'abri. Tour annulé, on relance.`));
  }

  function selectUserForRun(runHistory, runImmune, runForcedQueue, ignoreSafe) {
    if (runForcedQueue.length > 0) {
      const forcedIndex = runForcedQueue.findIndex(user => !safeSet.has(user));
      if (forcedIndex !== -1) {
        const [forcedUser] = runForcedQueue.splice(forcedIndex, 1);
        return { selectedUser: forcedUser, restartAll: false };
      }
      if (!ignoreSafe && runForcedQueue.length > 0 && safeSet.has(runForcedQueue[0])) {
        return { selectedUser: runForcedQueue[0], restartAll: true, safeHit: true };
      }
    }

    const excludeSafe = ignoreSafe ? safe : [];
    const candidateAll = selectUserForChore(runHistory, { exclude: excludeSafe });
    if (!candidateAll) {
      return { selectedUser: null, restartAll: false };
    }

    if (!ignoreSafe && safeSet.has(candidateAll)) {
      return { selectedUser: candidateAll, restartAll: true, safeHit: true };
    }

    if (runImmune.includes(candidateAll)) {
      const excludeList = [...new Set([...runImmune, ...safe])];
      const alternative = selectUserForChore(runHistory, { exclude: excludeList });
      if (alternative) {
        const immuneIndex = runImmune.indexOf(candidateAll);
        if (immuneIndex !== -1) {
          runImmune.splice(immuneIndex, 1);
        }
        if (!runForcedQueue.includes(candidateAll)) {
          runForcedQueue.push(candidateAll);
        }
        console.log(chalk.gray(`🛡️ Totem actif : ${candidateAll} reporté au prochain tour.`));
        return { selectedUser: alternative, restartAll: false };
      }
    }

    return { selectedUser: candidateAll, restartAll: false };
  }

  let ignoreSafe = false;
  let attempts = 0;
  const maxAttempts = 5;

  while (attempts < maxAttempts) {
    attempts += 1;
    const runHistory = { ...history };
    const runAssignments = [];
    const runImmune = [...immune];
    const runForcedQueue = [...forcedQueue];
    let restartAll = false;

    for (const chore of choresToAssign) {
      const result = selectUserForRun(runHistory, runImmune, runForcedQueue, ignoreSafe);
      if (result.restartAll) {
        if (result.safeHit && result.selectedUser) {
          await spinWheel(users, result.selectedUser, chore);
          showSafeArt(result.selectedUser);
        }
        restartAll = true;
        ignoreSafe = true;
        break;
      }

      const selectedUser = result.selectedUser;
      if (!selectedUser) {
        console.log(chalk.red('\n❌ Impossible de sélectionner un utilisateur non protégé.\n'));
        return;
      }

      runHistory[selectedUser]++;
      runAssignments.push({
        date: today,
        chore: chore,
        user: selectedUser,
        completed: false
      });
    }

    if (restartAll) {
      continue;
    }

    for (const assignment of runAssignments) {
      await spinWheel(users, assignment.user, assignment.chore);
      todayAssignments.push(assignment);
      assignments.push(assignment);
    }

    Object.assign(history, runHistory);
    immune = runImmune;
    forcedQueue = runForcedQueue;
    break;
  }

  if (attempts >= maxAttempts) {
    console.log(chalk.red('\n❌ Trop de relances dues aux utilisateurs "à l\'abri".\n'));
    return;
  }
  
  // Sauvegarder les données
  saveJSON(HISTORY_PATH, history);
  saveJSON(ASSIGNMENTS_PATH, assignments);
  saveJSON(TOTEM_PATH, { immune, forcedQueue, safe });
  
  // Message final spectaculaire
  console.log('');
  console.log(gradient.rainbow('═'.repeat(50)));
  console.log(gradient.pastel.multiline('✨ Répartition terminée avec succès ! ✨'));
  console.log(gradient.rainbow('═'.repeat(50)));
  console.log('');
}

module.exports = {
  assignChores
};
